Data entry and dashboard integration

adrianq opened the pull request:
### :pushpin: References

* **Issue:** Closes #89

### :memo: Implementation

- Two iframes with a bit of hacking to customise the styling and selection in data entry

TODO: remove non-valid periods and extend test coverage

### :art: Screenshots
![Screenshot from 2019-04-12 08-25-26](https://user-images.githubusercontent.com/6850223/56019646-1cbadb80-5cfd-11e9-8598-deccbf378eb0.png)
![Screenshot from 2019-04-12 08-25-47](https://user-images.githubusercontent.com/6850223/56019647-1cbadb80-5cfd-11e9-9264-c1a0179c7a62.png)

### :fire: Is there anything the reviewer should know to test it?
Install it as a regular DHIS2 app otherwise, you will have problems with the iframe origin policy


### :bookmark_tabs: Others

- Any change in the [metadata repo](https://github.com/EyeSeeTea/vaccination-metadata)? If so, what branch/PR?
- Any change in the [GUI library](https://github.com/EyeSeeTea/d2-ui-components)? If so, what branch/PR?

adrianq commented:
@tokland Ready for review!

- I will take care of the periods and help as part of this card https://github.com/EyeSeeTea/vaccination-app/issues/102
- I have created another card to keep in mind we need to restore the tests for the list https://github.com/EyeSeeTea/vaccination-app/issues/101

tokland commented on src/components/dashboard/Dashboard.jsx:
```diff
@@ -0,0 +1,105 @@
+        return (
+            <React.Fragment>
+                <PageHeader
+                    title={i18n.t("Dashboard")}
```
Makes sense to add the name of the campaign in the title? Dashboard: My test campaign. Same for Data Entry.

adrianq replied:
It is already present in the iframe... Anyway we can discuss the layout details when we take care of the refinement card.

tokland commented on src/components/data-entry/DataEntry.jsx:
```diff
@@ -0,0 +1,106 @@
+    };
+
+    state = {
+        isDataEntryIdValid: true,
```
Maybe start as false and set to true when loaded?

adrianq commented on src/components/campaign-configurator/CampaignConfigurator.jsx:
```diff
@@ -73,14 +72,16 @@ class CampaignConfigurator extends React.Component {
         {
             name: "dataEntry",
             icon: "library_books",
+            // TODO: isActive: (d2, dataSet) => canUpdate(d2, d2.models.dataSet, [dataSet]),
```
Yep, updated card https://github.com/EyeSeeTea/vaccination-app/issues/102

adrianq commented on src/models/datasets.js:
```diff
@@ -67,3 +67,23 @@ export function getDashboardId(dataSet, config) {
+        const fields = ["id,organisationUnits[id,name]"].join(",");
+        const dataSet = await d2.models.dataSets.get(id, { fields });
+        return dataSet.organisationUnits.toArray()[0].id;
+    } catch (err) {
```
I am doing this to check the id (dataset id or organisation id) is valid and display the snack bar if there is an error here https://github.com/EyeSeeTea/vaccination-app/pull/98/files#diff-5c15787cb2aa9317cc0fd63b4ca8cfd4R26 Better to move the try/catch to the parent function?

tokland replied:
An explicit snippet could be:

```
const orgUnits = dataSet.organisationUnits.toArray();
return _(orgUnits).isEmpty() ? undefined : orgUnits[0].id;
```

adrianq replied:
mmm but the problem is the 'get' not accessing the array
![image](https://user-images.githubusercontent.com/6850223/56032087-1d616b00-5d19-11e9-9d72-fb7ff8460be3.png)

tokland replied:
Sorry, I was focusing in the getting of the orgUnits. For the non-existing dataset this should work: `const dataSet = await d2.models.dataSets.get(id, { fields }).catch(err => undefined);` And then check if you have dataSet or not in the next line.

adrianq replied:
Nice one. Done https://github.com/EyeSeeTea/vaccination-app/pull/98/commits/83ab2176f1648181f9ec046073ada9556ea22a47

adrianq commented on src/components/data-entry/DataEntry.jsx:
```diff
@@ -0,0 +1,106 @@
+        iframeDocument.querySelector(`#leftBar`).style.top = "-10px";
+        iframeDocument.querySelector(`body`).style.marginTop = "-55px";
+        iframeDocument.querySelector(`#moduleHeader`).remove();
+    }
```
Yep, card update: https://github.com/EyeSeeTea/vaccination-app/issues/102

adrianq commented:
All the comments should be resolved now. Btw, I removed the componentDidUpdate and changed it for this approach https://github.com/EyeSeeTea/vaccination-app/pull/98/commits/b405ea336c1b8f04fc111142b2055068d69df641#diff-4f40ff551214939ddd69ed290990568aR22 I think it looks clearer but let me know...

tokland commented:
>  removed the componentDidUpdate and changed it for this approach 

All good, replacing a DidUpdate by a setState callback is idiomatic.
